polyfills/tcpsocket/tcpsocket.js

```diff
@@ -0,0 +1,309 @@
+    }
+
+    var host = extraData.host;
+    var port = extraData.port;
+    var options = extraData.options;
+
+    var _resolve, _reject;
+    var _sock = new Promise((resolve, reject) => {
+      _resolve = resolve;
+      _reject = reject;
+    });
+
+    // This will hold the remote object, once the lock is actually created..
+    var _sockId = null;
+    _sock.then(id => _sockId = id);
+
+    Object.defineProperty(this, 'host', {
+      get: function() {
+        return host;
+      }
+    });
+
+    Object.defineProperty(this, 'port', {
+      get: function() {
+        return port;
+      }
+    });
+
+    var _internalProps = {
+      ssl: !!options.useSecureTransport,
+      bufferedAmount: 0,
+      readyState: "connecting",
+      binaryType: options.binaryType == "arraybuffer" ? "arraybuffer"
+                                                      : "string"
+    };
+
+    Object.keys(_internalProps).forEach(key =>
+      Object.defineProperty(this, key, {
+        get: function() {
+          return _internalProps[key];
+        }
+      })
+    );
+
+    var _handlers = {
+      onopen: null,
+      ondrain: null,
+      ondata: null,
+      onerror: null,
+      onclose: null
+    };
+
+    function _sendAPICall(commandObject) {
+      Promise.all([navConnPromise, _sock]).then(values => {
+        // Going to send the socketId always in this place...
+        commandObject.data.socketId = values[1];
+        values[0].sendObject(commandObject);
+      });
+    }
+
+    Object.keys(_handlers).forEach(handler =>
+      Object.defineProperty(this, handler, {
+        get: function() {
+          return _handlers[handler];
+        },
+        set: function(cb) {
+          _handlers[handler] = cb;
+          navConnPromise.queueDependentRequest({handler: handler, cb: cb},
+                                               HandlerSetRequest,
+                                               _sock, 'socketId');
+        }
+      })
+    );
+
+    this.serialize = function() {
+      return {
+        id: reqId,
+        data: {
+          operation: 'open',
+          params: [host, port, options]
+        },
+        processAnswer: function(answer) {
+          // This function will be invoked in two cases... when the socket is
+          // created and any time we have to update something...
+          if (_sockId === null) {
+            if (!answer.error) {
+              _resolve(answer.socketId);
+              _internalProps.readyState = 'open';
+            } else {
+              var permaFail = 'Error creating socket: ' + answer.error;
+              _internalProps.readyState = 'closed';
+              _reject(permaFail);
+            }
+            return;
+          }
+
+          // Ok, so if we're here then the remote side has detected some kind of
+          // change that we have to apply locally...
+          // TO-DO: Stick it for now
+          debug('FIX-ME! I got some message that I\'m not treating yet!');
```

Comment:
This is probably not needed for OpenWapp, but also as a note to future-we, we should finish this